sign_api/src/rpc/params/shared_types.rs

```diff
@@ -0,0 +1,174 @@
+//! https://specs.walletconnect.com/2.0/specs/clients/sign/data-structures
+
+use {
+    serde::{Deserialize, Serialize},
+    std::collections::BTreeSet,
+};
+
+/// Errors covering namespace validation errors.
+///
+/// https://specs.walletconnect.com/2.0/specs/clients/sign/namespaces
+/// and some additional variants.
+#[derive(Debug, thiserror::Error)]
+pub enum NamespaceError {
+    #[error("Required chains are not supported: {0}")]
+    UnsupportedChains(String),
+    #[error("Chains must not be empty")]
+    UnsupportedChainsEmpty,
+    #[error("Chains must be CAIP-2 compliant")]
+    UnsupportedChainsCaip2,
+    #[error("Chains must be defined in matching namespace")]
+    UnsupportedChainsNamespace,
+    #[error("Required extensions are not supported")]
+    UnsupportedExtensions,
+    #[error("Required events are not supported: {0}")]
+    UnsupportedEvents(String),
+    #[error("Required methods are not supported: {0}")]
+    UnsupportedMethods(String),
+    #[error("Required namespace is not supported: {0}")]
+    UnsupportedNamespace(String),
+    #[error("Namespace formatting must match CAIP-2")]
+    UnsupportedNamespaceKey,
+}
+
+impl NamespaceError {
+    pub fn error_code(&self) -> i32 {
+        match self {
+            Self::UnsupportedChains(..)
+            | Self::UnsupportedChainsEmpty
+            | Self::UnsupportedChainsCaip2
+            | Self::UnsupportedChainsNamespace => 5100,
+            Self::UnsupportedEvents(..)
+            | Self::UnsupportedExtensions
+            | Self::UnsupportedMethods(..)
+            | Self::UnsupportedNamespace(..)
+            | Self::UnsupportedNamespaceKey => 5104,
+        }
+    }
+}
+
+#[derive(Debug, Serialize, PartialEq, Eq, Deserialize, Clone, Default)]
+#[serde(rename_all = "camelCase")]
+pub struct Metadata {
+    pub description: String,
+    pub url: String,
+    pub icons: Vec<String>,
+    pub name: String,
+}
+
+#[derive(Debug, Serialize, PartialEq, Eq, Deserialize, Clone, Default)]
+pub struct Relay {
+    pub protocol: String,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    #[serde(default)]
+    pub data: Option<String>,
+}
+
+/// https://specs.walletconnect.com/2.0/specs/clients/sign/namespaces
+#[derive(Debug, Serialize, Eq, PartialEq, Deserialize, Clone, Default)]
+#[serde(rename_all = "camelCase")]
+pub struct Namespaces {
```

Comment:
```
#[derive(Debug, Serialize, Eq, PartialEq, Deserialize, Clone, Default)]
#[serde(rename_all = "camelCase")]
pub struct Namespaces(BTreeMap<String, Namespace>);
```
Makes it more flexible, easier to validate, and can match CAIP-2 field names.
https://specs.walletconnect.com/2.0/specs/clients/sign/namespaces

Reply:
Done.